pkg/build/graph/analysis/bc.go

```diff
@@ -160,6 +245,44 @@ func FindPendingTags(g osgraph.Graph, f osgraph.Namer) []osgraph.Marker {
 	return markers
 }
 
+// validImageStreamImage will cycle through the imageStream.Status.Tags.[]TagEvent.DockerImageReference and  determine whether an image with the hexadecimal image id
+// associated with an ImageStreamImage reference in fact exists in a given ImageStream; on return, this method returns a true if does exist, and as well as the hexadecimal image
+// id from the ImageStreamImage, as well as the appropriate message to add to the marker if the image was not found
+func validImageStreamImage(imageNode *imagegraph.ImageStreamImageNode, imageStream *imageapi.ImageStream) (bool, string, string) {
+	dockerImageReference, err := imageapi.ParseDockerImageReference(imageNode.Name)
+	if err == nil {
+		for _, tagEventList := range imageStream.Status.Tags {
+			for _, tagEvent := range tagEventList.Items {
+				if strings.Contains(tagEvent.DockerImageReference, dockerImageReference.ID) {
+					return true, dockerImageReference.ID, ""
+				}
+			}
+		}
+	}
+
+	// check the images stream to see if any import images are in flight or have failed
+	annotation, ok := imageStream.Annotations[imageapi.DockerImageRepositoryCheckAnnotation]
+	if !ok {
+		return false, dockerImageReference.ID, "import the image with hexadecimal ID %s into the image stream %s"
+	}
+
+	if checkTime, err := time.Parse(time.RFC3339, annotation); err == nil {
+		// this time based annotation is set by pkg/image/controller/controller.go whenever import/tag operations are performed; unless
+		// in the midst of an import/tag oprertion, it stays set and serves as a timestamp for when the last operation occurred;
+		// so we will check if the image stream has been updated "recently";
+		// in case it is a slow link to the remote repo, see if if the check annotation occured within the last 5 minutes; if so, consider that as potentially "in progress"
+		compareTime := checkTime.Add(5 * time.Minute)
+		currentTime, _ := time.Parse(time.RFC3339, unversioned.Now().UTC().Format(time.RFC3339))
+		if compareTime.Before(currentTime) {
+			return false, dockerImageReference.ID, "import the image with hexadecimal ID %s into the image stream %s"
+		} else {
```

Comment:
@gabemontero the common idiom in Go:

```go
// don't use this

if condition {
	// DoStuff()
	return something
} else {
	// DoOtherStuff()
	return anotherthing
}
```

becomes


```go
// use this

if condition {
	// DoStuff()
	return something
}
// DoOtherStuff()
return anotherthing
```

This is part of the "mantra" of keeping the indentation level to a minimum, and returning earlier on exceptional conditions.